bpo-39102: Increase Enum performance up to 10x times (3x average)

Bobronium opened the pull request:
# Increase Enum performance 
https://bugs.python.org/issue39102

Partial benchmark result, to see more, check [full benchmark result and source code](https://gist.github.com/Bobronium/94b5ca3e4098669d93a6127e70359307)
```py
ATTR_ACCESS:
Testing with 10000 repeats, result is average of 10 tests:

    >>> NewEnum.foo  # 0.6069349023270748 ms
    <NewEnum.foo: 1>

    >>> OldEnum.foo  # 4.5089948174334005 ms
    <OldEnum.foo: 1>

    >>> NewEnum.foo.value  # 0.965516420197984 ms
    1

    >>> OldEnum.foo.value  # 5.234090615261106 ms
    1

    >>> NewEnum.value.value  # 0.6496817059283957 ms
    3

    >>> OldEnum.value.value  # 19.718928336346387 ms
    3

    >>> try:     NewEnum.value.value = 'new' except: pass  # 4.842046525117963 ms
    AttributeError("Can't set attribute")

    >>> try:     OldEnum.value.value = 'new' except: pass  # 24.484108522222897 ms
    AttributeError("can't set attribute")


NewEnum: total: 7.0641796 ms, average: 1.1652198 ms (Fastest)
OldEnum: total: 53.9461223 ms, average: 10.3317085 ms, ~ x8.87 times slower than NewEnum 
```
## [Full benchmark result and source code](https://gist.github.com/Bobronium/94b5ca3e4098669d93a6127e70359307)

## Changes
- Optimized `Enum.__new__`
- Remove `EnumMeta.__getattr__`,
- Store `Enum.name` and `.value` in members `__dict__` for faster access
- Deprecate getting values from `Enum._name_` and `._value_`, should use public `.name` and `.value` now
- Replace `Enum._member_names_` with `._unique_member_map_` for faster lookups and iteration
- Replace `_EmumMeta._member_names` and `._last_values` with `.members` mapping (deprecation warning on using old attrs)
- Add `_str_`, `_repr_` and `_invert_` attrs to cache results of functions of the same dunder names (x35 speed boost) (948d3de)
- Various minor improvements
- Add support for direct setting and getting class attrs on `DynamicClassAttribute` without need to use slow `__getattr__`


<!-- issue-number: [bpo-39102](https://bugs.python.org/issue39102) -->
https://bugs.python.org/issue39102
<!-- /issue-number -->

Bobronium committed:
Added class attributes access trough DynamicClassProperty (first attempt)

Bobronium committed:
Added class attributes access trough DynamicClassProperty (first attempt)

Bobronium committed:
Replaced dynamic attributes Enum.name and .value to instance attrs

Bobronium committed:
Improve values check speed (e.g. Color(3))

Bobronium committed:
Replace tuples with sets for faster lookups

Bobronium committed:
Used cls variable instead if multiple self.__class__

Bobronium committed:
Merge remote-tracking branch 'origin/fast-enum' into fast-enum

# Conflicts:
#	Lib/enum.py

Bobronium committed:
Fix whitespace in types

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

Bobronium committed:
Use _unique_member_map_ instead of _member_names_ to store unique enum member

Bobronium committed:
Use members dict instead of _member_names and _last_values on _EnumDict

Bobronium committed:
Rename DynamicClassAttribute.__init__(name -> alias), add tests

Bobronium committed:
DynamicClassAttribute.class_attr_name -> alias

Bobronium committed:
Restore original formatting

Bobronium committed:
Fix whitespaces

Bobronium committed:
Use dict instead of list to check values in _create_pseudo_member_

auvipy commented on Lib/enum.py:
```diff
@@ -101,12 +100,29 @@ def __setitem__(self, key, value):
+            self.members[key] = value
         super().__setitem__(key, value)
 
+    @property
```
the deprecation part could be addressed in a separate commit.

Bobronium replied:
I don't quite understand how it should look like, though. Should I change anything now?

Bobronium committed:
Deprecate getting values through _value_ and _name_
Just realized that _value_ and _name_ attrs remained independent from
name and value and could be different. This commit fixes it.

Bobronium committed:
Add 'name' and 'value' to Enum.__dir__

ethanfurman reviewed (commented):
Sorry a proper review is taking so long -- there is a lot to go over here.
What I can say so far:
- `name` and `value` must show up in a member's `dir()`;
- `_leading_underscore_names` and `plain_names` must not be used in `EnumMeta` nor `Enum` as then the user cannot use those same names -- that's why `_sunder_` and `__dunder__` names are used instead;

I do have some private code that I haven't merged in yet that does away with `DynamicClassAttribute` -- I'll try to get that done in the next few weeks.  I think a lot of the work you have done here will still be applicable.

Thank you for your efforts!

Bobronium committed:
Remove '_member_names' and return 'list(cls._unique_member_map_)' instead

Bobronium committed:
Use builtin dict() instead of _EnumDict for class creation

Bobronium committed:
Use f-strings instead of %s formatting

Bobronium committed:
Fix missing '>' in repr, add test for that

Bobronium commented on Lib/enum.py:
```diff
@@ -369,7 +366,7 @@ def __members__(cls):
 
     def __repr__(cls):
-        return "<enum %r>" % cls.__name__
+        return f'<enum {cls.__name__!r}'
```
Thanks! Fixed and added test for that.

Bobronium committed:
Add cache for repr, str and invert on Flag and IntFlag

Benchmark:
Testing with 10000 repeats, result is average of 100 tests:

    >>> str(~NewFlag.baz)  # 5.6313761773697415 ms
    'NewFlag.0'

    >>> str(~OldFlag.baz)  # 146.97604789150913 ms
    'OldFlag.0'

    >>> repr(~NewFlag.baz)  # 4.422742834372443 ms
    '<NewFlag.0: 0>'

    >>> repr(~OldFlag.baz)  # 151.49518529891247 ms
    '<OldFlag.0: 0>'

    >>> ~(~NewFlag.foo)  # 3.465736084655711 ms
    <NewFlag.foo: 1>

    >>> ~(~OldFlag.foo)  # 177.88357201820338 ms
    <OldFlag.foo: 1>

NewFlag: total: 13.5198551 ms, average: 4.4194400 ms (Fastest)
OldFlag: total: 476.3548052 ms, average: 158.2196475 ms, ~ x35.80 times slower than NewFlag

Bobronium committed:
Remove redundant name and value setattr

Bobronium commented on Lib/re.py:
```diff
@@ -153,17 +153,17 @@ class RegexFlag(enum.IntFlag):
-        value = self._value_
+        if self.name is not None:
+            return f're.{self.name}'
+        value = self.value
```
Should use `_repr_` cache here

Bobronium commented on Lib/enum.py:
```diff
@@ -143,38 +160,42 @@ def __new__(metacls, cls, bases, classdict):
         # Reverse value->name map for hashable values.
         enum_class._value2member_map_ = {}
 
+        # used to speedup __str__, __repr__ and __invert__ calls when applicable
```
Despite the fact that this works pretty well and gives a great performance boost, I'm not sure that it was implemented right, so feedback is highly appreciable.

Bobronium commented on Lib/enum.py:
```diff
@@ -564,30 +595,67 @@ def __new__(cls, value):
+        # TODO: Maybe remove try/except block and setting __context__ in this case?
         try:
-            exc = None
             result = cls._missing_(value)
         except Exception as e:
-            exc = e
-            result = None
+            # Huge boost for standard enum
+            if cls._missing_ is Enum._missing_:
+                raise
+            else:
+                e.__context__ = ValueError(f'{value!r} is not a valid {cls.__qualname__}')
+                raise
```
With this try/except block code runs twice slower for non-enum values checks, then without it . 
Do we really need to bother checking all errors and add `ValueError` to context , knowing it will happen either in standard `Enum._missing_` or in user-defined code? In fact, `Flag._missing_` also raises same error as `Enum._missing_` and handled and rerised too.

Bobronium reviewed (commented):
Feedback needed

ethanfurman reviewed (changes requested):
For 3.10 and supporting `__init_subclass__` I've made several structural changes to the `enum` implementation.  In doing that I included many of the optimizations that you spoke of.  If you are not already in the ACK file I'll add you there in my next commit.

Improvements made since your PR:

- `enum.property`, a subclass of `DynamicClassAttribute`, is used : it avoids the slow `__getattr__` lookup
- `decompose` has been removed (with belm0's help)
- `_invert_` is cached

Things that likely will not change:

- `_name_` as an `enum.property`
- `_value_` as an `enum.property`
- `__member_map__` as the public API for accessing enum internals

Things I am open to:

- using `set()` instead of lists
- using internal `dict()`s instead of matching `list()`s to track names/values
- using `f-strings`
- caching `__member_map__` instead of creating a new one on each access

If you would like to make another PR based on current 3.10 code based on above comments I would be happy to review and commit it.  If you don't have time or the desire to tackle this again, you still have my thanks for the ideas and your help in the improvements that have made it in so far.
